### PR TITLE
refactor: ed25519 signature verifier option

### LINKED ISSUE #1458: Take interface{} as public key for signature verifier
See https://github.com/hyperledger/aries-framework-go/pull/1447#discussion_r392448914

### PR DESCRIPTION
**Pull request by kdimak**
closes #1458

Signed-off-by: Dima <dkinoshenko@gmail.com>

Follow-up issue to use Crypto for VC: https://github.com/hyperledger/aries-framework-go/issues/1416
Follow-up issue to use Crypto in JWS/JWT: https://github.com/hyperledger/aries-framework-go/issues/1267

### CODE REVIEW

File: pkg/doc/signature/ed25519signature2018/suite.go [hunk 1]
```diff
@@ -81,18 +93,28 @@ func (s *SignatureSuite) GetDigest(doc []byte) []byte {
...
+			return errors.New("ed25519: bad public key length")
+		}
+
+		verified := ed25519.Verify(pubKey.Value, doc, signature)
+		if !verified {
+			return errors.New("signature doesn't match")
+		}
+
+		return nil
+
```
Comment (Baha-sk) [inline pkg/doc/signature/ed25519signature2018/suite.go]: I assume `case *keyset.Handle:` is coming up in another PR?

Comment (kdimak) [reply pkg/doc/signature/ed25519signature2018/suite.go]: I am basing on a more generic Crypto API interface which takes `kh interface{}`. In this way, not only **tinkrypto** is supported but any other crypto implementation as well. That's my idea.

Comment (Baha-sk) [reply pkg/doc/signature/ed25519signature2018/suite.go]: I'm ok with generic SignatureSuite API.

Similarly to KMS and Crypto APIs, they offer generic functions. The default implementation is using Tink (`*keyset.Handle`) as argument types.

same here, there should be a case for Tink's *keyset.Handle to call aries-framwork-go's default Crypto.Verify() instead of directly calling de25519 with []byte key

Comment (kdimak) [reply pkg/doc/signature/ed25519signature2018/suite.go]: It is what is happening in a `default` of the `switch`: we use Crypto instance passed as a functional argument to verify a signature. I think there is no need to introduce a direct dependency on `tinkcrypto` here.

Comment (kdimak) [reply pkg/doc/signature/ed25519signature2018/suite.go]: OK, we agreed with @Baha that can use `localkms.PubKeyBytesToHandle()` to convert public key bytes to key handle and use it with crypto. This will require some time to change, I will push the solution later on today.

Comment (kdimak) [reply pkg/doc/signature/ed25519signature2018/suite.go]: I think we should also check a type of the crypto impl. It it's a `tinkcrypto`, which can convert public key bytes to `*keyset.Handle` and call crypto.Verify(). In contrast, if other crypto impl is used then it could not accept `*keyset.Handle`, in this case, we will get into trouble. What do you think? @Baha-sk  @troyronda

File: pkg/doc/signature/ed25519signature2018/suite.go [hunk 2]
```diff
@@ -81,18 +93,28 @@ func (s *SignatureSuite) GetDigest(doc []byte) []byte {
...
 // Verify will verify a signature.
-func (s *SignatureSuite) Verify(pubKey *verifier.PublicKey, doc, signature []byte) error {
-	// ed25519 panics if key size is wrong
-	if l := len(pubKey.Value); l != ed25519.PublicKeySize {
-		return errors.New("ed25519: bad public key length")
+func (s *SignatureSuite) Verify(pubKeyValue interface{}, doc, signature []byte) error {
+	switch pubKey := pubKeyValue.(type) {
+	case *verifier.PublicKey:
+		// ed25519 panics if key size is wrong
+		if l := len(pubKey.Value); l != ed25519.PublicKeySize {
```
Comment (soluchok) [inline pkg/doc/signature/ed25519signature2018/suite.go]: `l := len(pubKey.Value); l != ed25519.PublicKeySize` -> `len(pubKey.Value) != ed25519.PublicKeySize`

File: pkg/doc/signature/ed25519signature2018/suite.go (hunk 1 above)
Comment (Baha-sk) [reply pkg/doc/signature/ed25519signature2018/suite.go]: this is why `case *keyset.Handle:` will manage Tink wrapped keys, other implementations can support `case verify.PublicKey`

The API will be flexible with these two case conditions.

File: pkg/doc/signature/ed25519signature2018/suite_test.go [hunk 3]
```diff
@@ -42,37 +49,102 @@ func TestSignatureSuite_Sign(t *testing.T) {
...
+	pubKeyBytes, err := lKMS.ExportPubKeyBytes(kid)
+	require.NoError(t, err)
+
+	pubKey := &sverifier.PublicKey{
+		Type:  kmsapi.Ed25519Type,
+		Value: pubKeyBytes,
+	}
+
+	err = ss.Verify(pubKey, doc, docSig)
+	require.NoError(t, err)
```
Comment (kdimak) [inline pkg/doc/signature/ed25519signature2018/suite_test.go]: @Baha-sk the test is failing here saying that signature is invalid:
```
verify msg: verifier_factory: invalid signature
```
Do you have any idea why it's happening?

Comment (kdimak) [reply pkg/doc/signature/ed25519signature2018/suite_test.go]: Filed an issue regarding this: https://github.com/hyperledger/aries-framework-go/issues/1483
I disabled accepting a public key bytes input when verifying a signature via `LocalCryptoVerifier`.

**Comment by kdimak** [comment]
I added `verifier` concept for the Signature Suite. It is defined by functional option in the same way as `signer` before. There are two default implementations of the `verifier`:
1.  `LocalCryptoVerifier` uses crypto and Local KMS for signature verification. The latter should be used to convert public key bytes to key handle and pass it to tinkcrypto (i removed support of this because of https://github.com/hyperledger/aries-framework-go/issues/1483).
2. `PublicKeyVerifier` is an old verification way using public key bytes. Other packages that use signature verification (like `did`  or `verifiable`) now use this type of verifier (mainly in the tests).

File: pkg/doc/did/doc.go [hunk 4]
```diff
@@ -790,6 +795,16 @@ func (r *didKeyResolver) Resolve(id string) (*verifier.PublicKey, error) {
 	return nil, ErrKeyNotFound
 }
 
+// MapKeyTypeToKMS maps key type of DID doc to the KMS one.
+func MapKeyTypeToKMS(t string) (kms.KeyType, error) {
+	switch t {
+	case "Ed25519VerificationKey2018":
+		return kms.Ed25519Type, nil
```
Comment (troyronda) [inline pkg/doc/did/doc.go]: I'm a bit confused. What's the purpose of using interface{} above when we are mapping to definitions in kms anyways?

File: pkg/doc/signature/ed25519signature2018/suite.go [hunk 5]
```diff
@@ -36,6 +42,89 @@ type signer interface {
...
+		localKMS: localKMS,
+	}
+}
+
+// Verify will verify a signature.
+func (v *LocalCryptoVerifier) Verify(pubKeyValue interface{}, doc, signature []byte) error {
+	var keyHandle *keyset.Handle
+
+	switch pubKey := pubKeyValue.(type) {
+	case *keyset.Handle:
```
Comment (troyronda) [inline pkg/doc/signature/ed25519signature2018/suite.go]: I don't really understand this change. 
It seems like we are using an interface{} and then requiring deep knowledge of the types (coupling the kms types).

ed25519 package should not know about tink.

File: pkg/doc/signature/ed25519signature2018/suite.go [hunk 6]
```diff
@@ -36,6 +42,89 @@ type signer interface {
...
+	return v.cr.Verify(signature, doc, keyHandle)
+}
+
+// PublicKeyVerifier verifies a Ed25519 signature taking Ed25519 public key bytes as input.
+type PublicKeyVerifier struct {
+}
+
+// Verify will verify a signature.
+func (v *PublicKeyVerifier) Verify(pubKeyValue interface{}, doc, signature []byte) error {
+	pubKey, ok := pubKeyValue.(*sigverifier.PublicKey)
```
Comment (troyronda) [inline pkg/doc/signature/ed25519signature2018/suite.go]: I'm a bit confused. It seems like we require a particular interface in this line, but not in the func signature?

File: pkg/doc/signature/ed25519signature2018/suite.go [hunk 7]
```diff
@@ -36,6 +42,89 @@ type signer interface {
...
+	if pubKey.Type != kms.Ed25519Type {
+		return fmt.Errorf("invalid public key type: %s", pubKey.Type)
+	}
+
+	// ed25519 panics if key size is wrong
+	if len(pubKey.Value) != ed25519.PublicKeySize {
+		return errors.New("ed25519: bad public key length")
+	}
+
+	verified := ed25519.Verify(pubKey.Value, doc, signature)
```
Comment (troyronda) [inline pkg/doc/signature/ed25519signature2018/suite.go]: this should be part of the kms/crypto implementation @Baha-sk

File: pkg/doc/signature/ed25519signature2018/suite.go [hunk 8]
```diff
@@ -36,6 +42,89 @@ type signer interface {
...
+}
+
+type verifier interface {
+	// Verify will verify a signature.
+	Verify(pubKeyValue interface{}, doc, signature []byte) error
+}
+
+// LocalCryptoVerifier defines a verifier which is based on Local KMS and Crypto
+// which uses keyset.Handle as input for verification.
+type LocalCryptoVerifier struct {
```
Comment (troyronda) [inline pkg/doc/signature/ed25519signature2018/suite.go]: This doesn't belong in this package @Baha-sk @kdimak

File: pkg/doc/did/doc.go [hunk 9]
```diff
@@ -747,7 +747,7 @@ type signatureSuite interface {
 	GetDigest(doc []byte) []byte
 
 	// Verify will verify signature against public key
-	Verify(pubKey *verifier.PublicKey, doc []byte, signature []byte) error
+	Verify(pubKey interface{}, doc []byte, signature []byte) error
```
Comment (troyronda) [inline pkg/doc/did/doc.go]: this change *only* makes sense if pubKey is an opaque object to every package except for a kms implementation.

File: pkg/doc/did/doc.go [hunk 10]
```diff
@@ -777,11 +777,16 @@ type didKeyResolver struct {
 	PubKeys []PublicKey
 }
 
-func (r *didKeyResolver) Resolve(id string) (*verifier.PublicKey, error) {
+func (r *didKeyResolver) Resolve(id string) (interface{}, error) {
```
Comment (troyronda) [inline pkg/doc/did/doc.go]: this change __only__ makes sense if pubKey is an opaque object to every package except for a kms implementation.

but I see this package constructing a generic return value. this is confusing.

File: pkg/doc/did/doc.go (hunk 9 above)
Comment (troyronda) [reply pkg/doc/did/doc.go]: and I also see resolve actually constructing a generic value... (so pubkey doesn't seem to be opaque at all).

File: pkg/doc/signature/ed25519signature2018/suite.go (hunk 6 above)
Comment (Baha-sk) [reply pkg/doc/signature/ed25519signature2018/suite.go]: `PublicKeyVerifier` is for keys in []byte format

File: pkg/doc/signature/ed25519signature2018/suite.go (hunk 5 above)
Comment (Baha-sk) [reply pkg/doc/signature/ed25519signature2018/suite.go]: this is because `publicKeyValue` can be either a key reference (Tink's *keyset.Handle) or the raw []byte publicKey, similar to Crypto API that takes a `kh interface` argument for the key reference and the default implementation uses Tink's *keyset.Handle as kh type

File: pkg/doc/signature/ed25519signature2018/suite.go (hunk 8 above)
Comment (Baha-sk) [reply pkg/doc/signature/ed25519signature2018/suite.go]: Agree, this struct can do any signature verification, not just ed25519.

ed25519signaturesuite2018 can use this struct to do verification, as well as any NIST approved ecdsa signature suite.

File: pkg/doc/signature/ed25519signature2018/suite.go (hunk 7 above)
Comment (Baha-sk) [reply pkg/doc/signature/ed25519signature2018/suite.go]: I think @kdimak wants to separate signature verification using raw key []byte from Tink keys and this verify uses []byte keys with crypto.ed25519 package.

So `PublicKeyVerifier` is for keys with raw bytes and `LocalKeyVerifier` is for Tink keys (with public key loaded from a DID doc as []byte and transformed by localkKMS into *keyset.Handle which requires a fix I'm currently working on)

File: pkg/doc/signature/ed25519signature2018/suite.go (hunk 8 above)
Comment (troyronda) [reply pkg/doc/signature/ed25519signature2018/suite.go]: More specifically, this package should not know what kms is in use nor anything about tink.

File: pkg/doc/signature/ed25519signature2018/suite_test.go (hunk 3 above)
Comment (Baha-sk) [reply pkg/doc/signature/ed25519signature2018/suite_test.go]: yeah, the signature has a special Tink prefix based on Tink key types, we can either fix it in Tink or manually remove the prefix from docSig and try `ss.Verify()` again... 

I will post an issue in Tink to see if we can get this added to Tink

File: pkg/doc/signature/ed25519signature2018/suite.go (hunk 7 above)
Comment (troyronda) [reply pkg/doc/signature/ed25519signature2018/suite.go]: This package should not know such details (that would be the reason for an empty interface).

File: pkg/doc/did/doc.go [hunk 11]
```diff
@@ -790,6 +795,16 @@ func (r *didKeyResolver) Resolve(id string) (*verifier.PublicKey, error) {
 	return nil, ErrKeyNotFound
 }
 
+// MapKeyTypeToKMS maps key type of DID doc to the KMS one.
+func MapKeyTypeToKMS(t string) (kms.KeyType, error) {
```
Comment (troyronda) [inline pkg/doc/did/doc.go]: This function doesn't belong in the did package.

**Comment by kdimak** [comment]
Now there is no dependency on KMS/Crypto in `ed25519signature2018`.  `keyResolver` produces generic `interface{}` and passes it to the `verifier`. I left `PublicKeyVerifier` in the production code as it is still used in many places in aries. Also, crypto operations like `ed25519.Verify()` are still present as I don't see a way now of how to use Crypto directly for this purpose. E.g. I would like to have the possibility to call a Crypto function with the following definition: 
```
Verify(signature, msg []byte, kms.KeyType type, privKey []byte)
```
If present, I would call it `Verify(signature, msg, kms.Ed25519Type, privKey)` and that's it.
I also added naive implementation of `signer/verifier` based on `crypto.Crypto` in a separate file. In the tests, one can see how it can be implemented using tinkcrypto/localkms.

File: pkg/doc/did/doc.go (hunk 11 above)
Comment (kdimak) [reply pkg/doc/did/doc.go]: Why? The key types are defined at https://www.w3.org/TR/did-core/#public-keys spec, and we need to map them to our internal key types.

File: pkg/doc/signature/ed25519signature2018/public_key_verifier.go [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+	"github.com/hyperledger/aries-framework-go/pkg/kms"
+)
+
+// PublicKeyVerifier verifies a Ed25519 signature taking Ed25519 public key bytes as input.
+type PublicKeyVerifier struct {
+}
+
+// Verify will verify a signature.
+func (v *PublicKeyVerifier) Verify(pubKeyValue interface{}, doc, signature []byte) error {
+	pubKey, ok := pubKeyValue.(*sigverifier.PublicKey)
```
Comment (troyronda) [inline pkg/doc/signature/ed25519signature2018/public_key_verifier.go]: This function should be calling a KMS implementation, not converting interfaces here... Is that the plan?

File: pkg/doc/signature/ed25519signature2018/suite_crypto.go [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+	return s.cr.Sign(msg, s.kh)
+}
+
+// CryptoVerifier defines signature verifier based on crypto.
+type CryptoVerifier struct {
+	cr crypto.Crypto
+}
+
+// NewCryptoVerifier creates a new CryptoVerifier.
+func NewCryptoVerifier(cr crypto.Crypto) *CryptoVerifier {
```
Comment (troyronda) [inline pkg/doc/signature/ed25519signature2018/suite_crypto.go]: Why is this only called by tests?

File: pkg/doc/signature/ed25519signature2018/public_key_verifier.go (hunk 12 above)
Comment (kdimak) [reply pkg/doc/signature/ed25519signature2018/public_key_verifier.go]: `NewCryptoVerifier()` is what we would like users to use. `PublicKeyVerifier` is a sort-of backward compatibility struct which is used in other packages (`verifiable`, `did`, `didexchange`,  `legacykms`) and should be replaced in the follow-up PRs. I can leave a comment regarding this in the code.

File: pkg/doc/signature/ed25519signature2018/suite_crypto.go (hunk 13 above)
Comment (kdimak) [reply pkg/doc/signature/ed25519signature2018/suite_crypto.go]: This will be changed in the follow-up PRs in order to not make the current PR that huge.

File: pkg/doc/did/doc.go (hunk 10 above)
Comment (sandrask) [reply pkg/doc/did/doc.go]: I agree that we should not return interface here.

File: pkg/doc/did/doc.go [hunk 14]
```diff
@@ -777,11 +777,16 @@ type didKeyResolver struct {
 	PubKeys []PublicKey
 }
 
-func (r *didKeyResolver) Resolve(id string) (*verifier.PublicKey, error) {
+func (r *didKeyResolver) Resolve(id string) (interface{}, error) {
 	for _, key := range r.PubKeys {
 		if key.ID == id {
+			keyType, err := MapKeyTypeToKMS(key.Type)
```
Comment (sandrask) [inline pkg/doc/did/doc.go]: We could defer key type  mapping when it is needed (sign, verify) and not tie it to resolution. 
Did package should not know about KMS types. This should be postponed until sign or verify. You can have helper function on crypto before PubKeyBytesToHandle(pubKey.Value, pubKey.Type) is called.

File: pkg/doc/signature/ed25519signature2018/suite_crypto_test.go [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+type Crypto struct {
+	*tinkcrypto.Crypto
+	localKMS *localkms.LocalKMS
+}
+
+func (t *Crypto) Verify(sig, msg []byte, kh interface{}) error {
+	var keyHandle *keyset.Handle
+
+	switch pubKey := kh.(type) {
+	case *keyset.Handle:
```
Comment (sandrask) [inline pkg/doc/signature/ed25519signature2018/suite_crypto_test.go]: I am not sure in what case handle would be passed in. Normally use case:
1. retrieve key from did (pub key bytes, type) 
2. use pub key for verification.

Comment (Baha-sk) [reply pkg/doc/signature/ed25519signature2018/suite_crypto_test.go]: usually pub key bytes will be transformed into public keyset.Handle to use kms/crypto's default implementation services

File: pkg/didcomm/protocol/didexchange/states.go [hunk 16]
```diff
@@ -591,10 +590,11 @@ func verifySignature(connSignature *ConnectionSignature, recipientKeys string) (
 	pubKey := base58.Decode(recipientKeys)
 
 	// TODO: Replace with signed attachments issue-626
-	signatureSuite := ed25519signature2018.New()
+	suiteVerifier := &ed25519signature2018.PublicKeyVerifier{}
+	signatureSuite := ed25519signature2018.New(ed25519signature2018.WithVerifier(suiteVerifier))
 
 	err = signatureSuite.Verify(&verifier.PublicKey{
-		Type:  kms.Ed25519Type,
+		Type:  "ED25519",
```
Comment (troyronda) [inline pkg/didcomm/protocol/didexchange/states.go]: why this change?

File: pkg/doc/did/doc.go (hunk 14 above)
Comment (kdimak) [reply pkg/doc/did/doc.go]: I agree with you and @troyronda . I removed this conversion. I also changed the `Type` of the `PublicKey` to be just a string as we discussed on the call with @sandrask .

File: pkg/didcomm/protocol/didexchange/states.go (hunk 16 above)
Comment (sandrask) [reply pkg/didcomm/protocol/didexchange/states.go]: I commented that we tie kms to our signature suite. Now that I see this it doesn't look good either. Dima please revert back. I expected this type to be passed in from resolver - not hard-coded.

File: pkg/doc/signature/ed25519signature2018/suite_crypto_test.go (hunk 15 above)
Comment (kdimak) [reply pkg/doc/signature/ed25519signature2018/suite_crypto_test.go]: I changed `publicKeyResolver.Resolve()` and `verifier.Verify()` to take `PublicKey` instead of genetic `interface{}`. The reason is that we don't see any use case to take e.g.  `*keyset.Handle` for now.
It looks like that when we would like to use `tinkcypto.Verifier`, we can intercept the call and convert `PublicKey` into the `*keyset.Handle` using `PubKeyBytesToHandle()`. This approach is implemented in **suite_crypto_test.go**.

File: pkg/didcomm/protocol/didexchange/states.go (hunk 16 above)
Comment (kdimak) [reply pkg/didcomm/protocol/didexchange/states.go]: The type is now a `string`, not `kms.KeyType`. I decided it will look simpler in the tests to use just constant instead of `string(kms.Ed25519Type)` (plus extra import).

Comment (kdimak) [reply pkg/didcomm/protocol/didexchange/states.go]: I updated constants usage. As @troyronda suggested, I added string key type constants to the KMS and re-used it for KeyType constants.

Comment (sandrask) [reply pkg/didcomm/protocol/didexchange/states.go]: sounds good

File: pkg/doc/signature/ed25519signature2018/public_key_verifier.go [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+// NOTE: this verifier is present for backward compatibility reasons and can be removed soon.
+// Please use CryptoVerifier or your own verifier.
+type PublicKeyVerifier struct {
+}
+
+// Verify will verify a signature.
+func (v *PublicKeyVerifier) Verify(pubKey *sigverifier.PublicKey, doc, signature []byte) error {
+	// ed25519 panics if key size is wrong
+	if len(pubKey.Value) != ed25519.PublicKeySize {
+		return errors.New("ed25519: bad public key length")
```
Comment (Baha-sk) [inline pkg/doc/signature/ed25519signature2018/public_key_verifier.go]: change "ed25519: bad public key length" to "ed25519: invalid key"

File: pkg/doc/signature/ed25519signature2018/public_key_verifier.go [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+	}
+
+	verified := ed25519.Verify(pubKey.Value, doc, signature)
+	if !verified {
+		return errors.New("signature doesn't match")
```
Comment (Baha-sk) [inline pkg/doc/signature/ed25519signature2018/public_key_verifier.go]: change "signature doesn't match" with "ed25519: invalid signature"

File: pkg/doc/signature/ed25519signature2018/public_key_verifier.go (hunk 17 above)
Comment (kdimak) [reply pkg/doc/signature/ed25519signature2018/public_key_verifier.go]: changed

File: pkg/doc/signature/ed25519signature2018/public_key_verifier.go (hunk 18 above)
Comment (kdimak) [reply pkg/doc/signature/ed25519signature2018/public_key_verifier.go]: changed
